app/src/test/kotlin/com/wire/android/notification/WireNotificationManagerTest.kt

```diff
@@ -238,16 +248,35 @@ class WireNotificationManagerTest {
                 .withCurrentScreen(CurrentScreen.InBackground)
                 .arrange()
 
-            manager.fetchAndShowNotificationsOnce(userId.value)
-            manager.fetchAndShowNotificationsOnce(userId.value)
-            runCurrent()
+            coEvery { arrangement.connectionPolicyManager.handleConnectionOnPushNotification(userId) } coAnswers {
+                // Push handling is taking 10 minutes
+                delay(10.minutes)
+            }
+            launch {
+                // We call fetchAndShow now
+                manager.fetchAndShowNotificationsOnce(userId.value)
+            }
+            launch {
+                // After 5 minutes we call fetchAndShow again
+                delay(5.minutes)
+                manager.fetchAndShowNotificationsOnce(userId.value)
+            }
+            // After first call, should have handled push once
+            advanceTimeBy(1.minutes.inWholeMilliseconds)
+            coVerify(exactly = 1) { arrangement.connectionPolicyManager.handleConnectionOnPushNotification(userId) }
+
+            // After second call, should have handled push once
+            advanceTimeBy(6.minutes.inWholeMilliseconds)
+            coVerify(exactly = 1) { arrangement.connectionPolicyManager.handleConnectionOnPushNotification(userId) }
```

Comment:
Keep in mind that these delays do not delay the execution of the tests.

Because we're calling `runTest` with a `TestDispatcher`, these delays will just wait until the virtual clock is advanced, or things calls to `runCurrent` to advance the next delay, or `advanceUntilIdle` to run through all child jobs until nothing else remains.

Reply:
Just one question: in tests we are not able to test doze mode and were connection breaks because this depends on the device? If yes, I think we should definitely think about some integration test on some real devices which can run once per week 🤔

Reply:
I know we can run `adb` commands to test doze mode, etc.

But considering that Instrumentation Tests kinda rely on the phone being connected, responsive and executing a lot of stuff, I'm not sure it's something easy to do.
Considering that Instrumented Tests kinda put the app on the foreground, Doze mode probably is off when it all happens.
On a quick search I wasn't able to find ways to do it.

It seems more like an end-to-end test with devices. Something that maybe QA could do? It would be something flaky I guess 🤔